src/resolve/resolved-dns-packet.c

```diff
@@ -2570,6 +2570,67 @@ bool dns_packet_equal(const DnsPacket *a, const DnsPacket *b) {
         return dns_packet_compare_func(a, b) == 0;
 }
 
+int dns_packet_ede_rcode(DnsPacket *p, uint16_t *ret_ede_rcode) {
+        assert(p);
+        bool has_ede = false;
+        uint16_t ede_rcode = _DNS_EDNS_OPT_MAX_DEFINED;
+        const uint8_t *d;
+        size_t l;
+
+        if (!p->opt)
+                return false;
+
+        d = p->opt->opt.data;
+        l = p->opt->opt.data_size;
+
+        while (l > 0) {
+                uint16_t code, length;
+
+                if (l < 4U)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                               "EDNS0 variable part has invalid size.");
+
+                code = unaligned_read_be16(d);
+                length = unaligned_read_be16(d + 2);
+
+                if (l < 4U + length)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                               "Truncated option in EDNS0 variable part.");
+
+                if (code == DNS_EDNS_OPT_EXT_ERROR) {
+                        if (length < 2U)
+                                return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                                "EDNS0 truncated EDE info code.");
+                        has_ede = true;
+                        ede_rcode = unaligned_read_be16(d+4);
+                        break;
+                }
+
+                d += 4U + length;
+                l -= 4U + length;
+        }
+
+        if (has_ede && ede_rcode < _DNS_EDNS_OPT_MAX_DEFINED)
+                *ret_ede_rcode = ede_rcode;
```

Comment:
we have this rule in systemd that in functions on success all return params are initialized and on failure none are touched. (where success is defined as the function returning >= 0 and  failure as returning < 0). Here you update `*ret_ede_rcode` only in some cases of success. 

please make sure to set the return param to something resonable in case the EDE data is not available. (i.e. UINT16_MAX or so?)

Reply:
Ok, PTAL. Now returns -EINVAL if we don't understand the code too.